### PR TITLE
create configuration cli

### PR DESCRIPTION
**Pull request by StuartHarris**
- [x] use a master yaml file that describes public services in their stacks and `.dev` TLD
- [x] assign unused ports above 8000 to services that are not yet in the swarm
- [x] generate an `nginx` config to do `host`-header-based routing
- [x] use `dnsmasq` to provide local dns resolution for `.dev` TLD to localhost
- [x] generate compose file overlay (to assign ports to published services)
- [x] update nginx configuration on deploy
- [x] include names of compose files in `stacks.yml`
- [x] only include ports for services that exist in compose files
- [ ] ~deploy a given stack~ (...moving this to https://github.com/redbadger/stack/issues/22)

### CODE REVIEW

File: provisioning/osx/configure/src/compose-file.js [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+};
+
+export const merge = async (mergeFn, dir, filesByStack) => {
+  const retVal = {};
+  for (const [stack, files] of R.toPairs(filesByStack)) {
+    const args = R.chain(f => ['-f', f], R.map(f => path.join(dir, f), files));
+    retVal[stack] = await mergeFn('docker-compose', [...args, 'config']);
+  }
+  return retVal;
+};
```
Comment (tabazevedo) [inline provisioning/osx/configure/src/compose-file.js:44]: Could this be written as a reducer? Or does the async/await transpiling blow up?

Off the top of my head something like...

```js
export const merge = async (mergeFn, dir, filesByStack) => R.reduce(
  async (accumulator, [stack, files]) => {
    const args = R.chain(f => ['-f', f], R.map(f => path.join(dir, f), files));
    const merged = await mergeFn('docker-compose', [...args, 'config']);
    return R.set(R.lensProp(stack), merged, accumulator);
  }, {},
  R.toPairs(filesByStack)
);
```

Don't mind me though... a `for..of` is probably fine 😄

File: provisioning/osx/configure/src/compose-file.js [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+export const writeFn = (filePath, content) => {
+  console.log(`Writing ${filePath}`); // eslint-disable-line
+  fs.writeFileSync(filePath, content);
+};
+
+export const write = (writeFn, filesByStack, dir, prefix) => {
+  const paths = {};
```
Comment (tabazevedo) [inline provisioning/osx/configure/src/compose-file.js:52]: Another good candidate for `R.reduce` ?

File: provisioning/osx/configure/src/index.js [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+import { getServices, getComposeFiles } from './config';
+import {
+  create as createNginxConfig,
+  reload as reloadNginx,
+  write as writeNginxConfig,
+} from './nginx';
+import { assign as assignPorts } from './ports';
+import { findWithPublishedPorts as findPublicServices } from './services';
+
+const dockerEnv = Bluebird.promisify(DockerMachine.env);
```
Comment (tabazevedo) [inline provisioning/osx/configure/src/index.js]: If we're on Node 8 here we can use https://nodejs.org/api/util.html#util_util_promisify_original and drop the `Bluebird` dependency

File: provisioning/osx/configure/src/nginx.js [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    return 404;
+  }
+}
+`;
+
+export const create = (services, domain) => `
+${baseConfig}${R.join(
+  '',
+  R.map(
+    s => `
```
Comment (tabazevedo) [inline provisioning/osx/configure/src/nginx.js:17]: A little hard to follow this, possibly pull out the string template into its own function?

File: provisioning/osx/configure/src/config.spec.js [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+        aliases: [],
+      },
+      {
+        stack: 'app',
+        name: 'rproxy',
+        aliases: ['web'],
+      },
+    ];
+    const actual = getServices(config);
+    expect(actual).to.deep.equal(expected);
```
Comment (tabazevedo) [inline provisioning/osx/configure/src/config.spec.js:33]: Just a forethought, I've grown to dislike the `expect/should` syntax as it makes it seem like natural language but it can trip you up.

I think they've fixed it in the more recent versions, but things like `expect(a).to.be.defined;` used to always pass as `.to.be.defined` was not a real assertion (even though `.to.be.undefined` was).

IIRC it now throws an error saying its not a valid property, but I still prefer the more declarative `assert` syntax where everything is a function.

File: readme.md [hunk 6]
```diff
@@ -60,19 +87,31 @@ Incoming requests can hit any node of the swarm and will be routed to an instanc
...
+And the following local containers running:
+
+```sh
+on-local docker ps
+CONTAINER ID        IMAGE               COMMAND                  CREATED             STATUS              PORTS                                    NAMES
+fbff0358c773        andyshinn/dnsmasq   "dnsmasq -k -d -A ..."   9 hours ago         Up 9 hours          0.0.0.0:53->53/tcp, 0.0.0.0:53->53/udp   dns_dns_1
+dc929dc479e0        nginx               "nginx -g 'daemon ..."   46 hours ago        Up 46 hours         0.0.0.0:80->80/tcp                       loadbalancer_load_balancer_1
+6e2686acd14a        registry:2          "/entrypoint.sh /e..."   46 hours ago        Up 46 hours         0.0.0.0:5001->5000/tcp                   registry_registry-mirror_1
+a578c7c8703c        registry:2          "/entrypoint.sh /e..."   46 hours ago        Up 46 hours         0.0.0.0:5000->5000/tcp                   registry_registry_1
 ```
```
Comment (tabazevedo) [inline readme.md:115]: This is so cool :scream:

**Review by tabazevedo — CHANGES_REQUESTED** [review]
Here's my 2 cents on this stuff so far. It's awesome! Could only comment from a code perspective as I haven't been keeping up with the functionality side of this, so you might want @charypar to also take a look

File: provisioning/osx/configure/src/compose-file.js (hunk 1 above)
Comment (StuartHarris) [reply provisioning/osx/configure/src/compose-file.js:44]: Hey, thanks for the review comments! Awesome.
I originally spent a while trying to do exactly this, but was struggling with the async reducer function. But I just tried your code and it works! Thanks. Much better.
(Get you with your lenses) :-)

Comment (StuartHarris) [reply provisioning/osx/configure/src/compose-file.js:44]: Hmm, not so fast Stupid Stuey. :-)
I tried it on the example below and it doesn't seem to wait for the promise to resolve so the test fails. I looked at the source code for `R.reduce` and there's no handling of async functions in there. :-( Would have been so nice though.

File: provisioning/osx/configure/src/compose-file.js (hunk 2 above)
Comment (StuartHarris) [reply provisioning/osx/configure/src/compose-file.js:52]: see above.
Also, there's a massive side effect (writing a file) so I wan't that to be obvious by using a `forEach`

File: provisioning/osx/configure/src/index.js (hunk 3 above)
Comment (StuartHarris) [reply provisioning/osx/configure/src/index.js]: Currently I'm transpiling with Babel to the user's node version so I'm guessing it'll get polyfilled. Anyway works for me (on node 8) so we'll find out soon enough on other people's machines. But good call. Thanks

File: provisioning/osx/configure/src/nginx.js (hunk 4 above)
Comment (StuartHarris) [reply provisioning/osx/configure/src/nginx.js:17]: Yeah I know it's a pile of mess. These things are never clean without using a proper templating library. Which I should probably have done. This file is deleted in https://github.com/redbadger/stack/pull/23 so won't bother to change

File: provisioning/osx/configure/src/config.spec.js (hunk 5 above)
Comment (StuartHarris) [reply provisioning/osx/configure/src/config.spec.js:33]: Yeah, I agree. I got into doing what your describing when using Ava and it's much better, but didn't really think about it this time and fell back to old habits. :-) Not going to change them all now, but will use the `assert` family in future.

File: provisioning/osx/configure/src/compose-file.js [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+services:
+${R.join('', R.map(genService, services))}
+`;
+
+  return R.fromPairs(
+    R.map(([stackname, services]) => [stackname, genStack(services)], stackNameAndServices),
+  );
+};
+
+// const deleteFile = ({ name }) => fs.unlinkSync(name);
```
Comment (charypar) [inline provisioning/osx/configure/src/compose-file.js]: Old code?

File: provisioning/osx/configure/src/config.spec.js (hunk 5 above)
Comment (charypar) [reply provisioning/osx/configure/src/config.spec.js:33]: FWIW, I still prefer the `expect`s :)

File: provisioning/osx/configure/src/services.js [hunk 8]
```diff
@@ -0,0 +1,12 @@
+import R from 'ramda';
+
+export const findWithPublishedPorts = R.pipe(
+  R.map(s => {
+    const name = s.Spec.Name.split('_')[1];
```
Comment (charypar) [inline provisioning/osx/configure/src/services.js]: What if a stack name has a `_` in it?

File: provisioning/osx/configure/src/services.spec.js [hunk 9]
```diff
@@ -0,0 +1,398 @@
+import { expect } from 'chai';
+import { findWithPublishedPorts } from './services';
+
+describe('services', () => {
+  it('should find public services', () => {
+    const services = [
```
Comment (charypar) [inline provisioning/osx/configure/src/services.spec.js:6]: This may be a little brittle for API versions other than the one you've tested against. Nothing to really do about it, just... FYI. It does change slightly between the dot versions.

**Review by charypar — CHANGES_REQUESTED** [review]
Only a couple minor comments, otherwise awesome!

**Review by tabazevedo — APPROVED** [review]
Happy with your comments and the changes you made after my review. I tried to do the reduce as well for a little while, got it working with [Bluebird.props](http://bluebirdjs.com/docs/api/promise.props.html) but it just ends up being more complex than the `for..of`s so lets leave it at that!

:+1:

**Review by charypar — APPROVED** [review]
